components/sensor/ina219/ina219.go

```diff
@@ -0,0 +1,236 @@
+		modelname,
+		registry.Component{Constructor: func(
+			ctx context.Context,
+			deps registry.Dependencies,
+			config config.Component,
+			logger golog.Logger,
+		) (interface{}, error) {
+			attr, ok := config.ConvertedAttributes.(*AttrConfig)
+			if !ok {
+				return nil, rdkutils.NewUnexpectedTypeError(AttrConfig{}, config.ConvertedAttributes)
+			}
+			return newSensor(ctx, deps, config.Name, attr, logger)
+		}})
+
+	config.RegisterComponentAttributeMapConverter(sensor.Subtype, modelname,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf AttrConfig
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		}, &AttrConfig{})
+}
+
+func newSensor(
+	ctx context.Context,
+	deps registry.Dependencies,
+	name string,
+	attr *AttrConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+	b, err := board.FromDependencies(deps, attr.Board)
+	if err != nil {
+		return nil, fmt.Errorf("ina219 init: failed to find board: %w", err)
+	}
+	localB, ok := b.(board.LocalBoard)
+	if !ok {
+		return nil, fmt.Errorf("board %s is not local", attr.Board)
+	}
+	i2cbus, ok := localB.I2CByName(attr.I2CBus)
+	if !ok {
+		return nil, fmt.Errorf("ina219 init: failed to find i2c bus %s", attr.I2CBus)
+	}
+	addr := attr.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Warnf("using i2c address : %d", defaultI2Caddr)
+	}
+
+	s := &ina219{
+		name:    name,
+		logger:  logger,
+		bus:     i2cbus,
+		busName: attr.I2CBus,
+		addr:    byte(addr),
+	}
+
+	err = s.calibrate(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// ina219 is a i2c sensor device that reports voltage, current and power
+type ina219 struct {
+	generic.Unimplemented
+	logger     golog.Logger
+	bus        board.I2C
+	busName    string
+	addr       byte
+	name       string
+	currentLSB int64
+	powerLSB   int64
+	cal        int64
+}
+
+type PowerMonitor struct {
+	Shunt   int64
+	Voltage float64
+	Current float64
+	Power   float64
+}
+
+func (d *ina219) calibrate(ctx context.Context) error {
+	if senseResistor <= 0 {
+		return fmt.Errorf("ina219 calibrate: senseResistor value invalid %d", senseResistor)
+	}
+	if maxCurrent <= 0 {
+		return fmt.Errorf("ina219 calibrate: maxCurrent value invalid %d", maxCurrent)
+	}
+
+	d.currentLSB = maxCurrent / (1 << 15)
+	d.powerLSB = int64((maxCurrent*20 + (1 << 14)) / (1 << 15))
+	// Calibration Register = 0.04096 / (current LSB * Shunt Resistance)
+	// Where lsb is in Amps and resistance is in ohms.
+	// Calibration register is 16 bits.
+	cal := calibratescale / (int64(d.currentLSB) * int64(senseResistor))
+	if cal >= (1 << 16) {
+		return fmt.Errorf("ina219 calibrate: calibration register value invalid %d", cal)
+	}
+	d.cal = cal
+
+	return nil
+}
+
+// Readings returns a list containing three items (voltage, current, and power).
+func (d *ina219) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	handle, err := d.bus.OpenHandle(d.addr)
```

Comment:
When you open an I2C bus handle, you lock the bus until your handle is closed, so nothing else can use it. Put a `defer handle.Close()` on line 173, so that even if we return an error, the bus is usable by someone else afterwards.

Reply:
done, thanks